### PR TITLE
feature(core): add ExecutionContext, ArgumentsHost type check #1581

### PR DESCRIPTION
**Pull request by kamilmysliwiec**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #1581

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information

### CODE REVIEW

File: packages/core/helpers/execution-context-host.ts [hunk 1]
```diff
@@ -1,18 +1,32 @@
...
+  private contextType: TContext = 'http' as TContext;
 
-export class ExecutionContextHost implements ExecutionContext {
   constructor(
     private readonly args: any[],
     private readonly constructorRef: Type<any> = null,
     private readonly handler: Function = null,
   ) {}
 
+  setType(type: TContext) {
```
Comment (BrunnerLivio) [inline packages/core/helpers/execution-context-host.ts]: Should `setType()` be part of the public API? It does not really do anything in perspective of a user.
Can't we just make the `contextType` part of the constructor and pass it internally?

File: packages/core/interceptors/interceptors-consumer.ts [hunk 2]
```diff
@@ -1,24 +1,30 @@
...
     args: any[],
     instance: Controller,
     callback: (...args: any[]) => any,
     next: () => Promise<any>,
+    type?: TContext,
   ): Promise<any> {
     if (isEmpty(interceptors)) {
       return next();
     }
     const context = this.createContext(args, instance, callback);
```
Comment (BrunnerLivio) [inline packages/core/interceptors/interceptors-consumer.ts]:
```suggestion
    const context = this.createContext(args, instance, callback, type);
    // Remove `context.setType(type)`
```

.. you get the idea.

File: packages/core/helpers/execution-context-host.ts (hunk 1 above)
Comment (kamilmysliwiec) [reply packages/core/helpers/execution-context-host.ts]: So the reason is that we need to allow setting type from some external libs (that are still a part of the ecosystem), like `@nestjs/graphql`

Comment (BrunnerLivio) [reply packages/core/helpers/execution-context-host.ts]: Makes sense..

Comment (fredtma) [reply packages/core/helpers/execution-context-host.ts]: You get an error If you use @nestjs/core 6.0.0 with @nestjs/websocket 6.8.2 bcos of this. websocket expect property `setType` but core does not have the property `setType`. [see](https://stackoverflow.com/questions/58317044/nestjs-websockets-with-guards-fails-with-host-settype/58318154#58318154)
I know it's from two different packages but it is a breaking change, unless you upgrade core to 6.8.2

**Pull request merged**